src/Prefetch.cpp

```diff
@@ -0,0 +1,280 @@
+            elem_size_bytes = elem_size_var;
+        }
+
+        std::ostringstream ss; ss << t;
+        string type_name = ss.str();
+        string pstr = unique_name('p');
+        debug(1) << "  prefetch #" << pstr << ": "
+                             << varname << " ("
+                             << (has_static_type ? type_name : elem_size_name)
+                             << ", dims:" << dims << ")\n";
+
+        // TODO: Opt: check box if it should be prefetched?
+        // TODO       - Only prefetch if varying by loop_var_name?
+        // TODO       - Don't prefetch if "small" all constant dimensions?
+        // TODO         e.g. see: camera_pipe.cpp corrected matrix(4,3)
+
+        string varname_prefetch_buf = varname + "_prefetch_" + pstr + "_buf";
+        Expr var_prefetch_buf = Variable::make(Int(32), varname_prefetch_buf);
+
+        // Establish the variables for buffer strides, box min & max
+        vector<Expr> stride_var(dims);
+        vector<Expr> extent_var(dims);
+        vector<Expr> min_var(dims);
+        vector<Expr> max_var(dims);
+        for (int i = 0; i < dims; i++) {
+            string istr = std::to_string(i);
+            // string extent_name = varname + ".extent." + istr;
+            string stride_name = varname + ".stride." + istr;
+            string extent_name = varname + ".extent." + istr;
+            string min_name = varname + "_prefetch_" + pstr + "_min_" + istr;
+            string max_name = varname + "_prefetch_" + pstr + "_max_" + istr;
+
+            stride_var[i] = Variable::make(Int(32), stride_name);
+            extent_var[i] = Variable::make(Int(32), extent_name);
+            min_var[i] = Variable::make(Int(32), min_name);
+            max_var[i] = Variable::make(Int(32), max_name);
+
+            // Record let assignments
+            // except for stride, already defined elsewhere
+            plets.push_back(make_pair(min_name, box[i].min));
+            plets.push_back(make_pair(max_name, box[i].max));
+        }
+
+        // Create a buffer_t object for this prefetch.
+        vector<Expr> args;
+
+        Expr first_elem = Load::make(t, varname, 0, BufferPtr(), Parameter());
+        args.push_back(Call::make(Handle(), Call::address_of, {first_elem}, Call::PureIntrinsic));
+        args.push_back(make_zero(t));
+        for (int i = 0; i < dims; i++) {
+            args.push_back(min_var[i]);
+            args.push_back(max_var[i] - min_var[i] + 1);
+            args.push_back(stride_var[i]);
+        }
+
+        // Create the create_buffer_t call
+        Expr prefetch_buf = Call::make(type_of<struct buffer_t *>(), Call::create_buffer_t,
+                              args, Call::Intrinsic);
+        plets.push_back(make_pair(varname_prefetch_buf, prefetch_buf));
+
+        // Create the prefetch call
+        Expr num_elem = stride_var[dims-1] * extent_var[dims-1];
+        vector<Expr> args_prefetch = {
+            dims,
+            elem_size_bytes,
```

Comment:
The elem_size_bytes is not correct in the constructed buffer_t when the type is not statically available but instead only available at runtime (e.g. as it the case with input) so I passed the element size to the prefetch runtime since the prefetch pass had already determined the correct value.